diff --git a/core/modules/filter/src/Plugin/Filter/FilterAutoP.php b/core/modules/filter/src/Plugin/Filter/FilterAutoP.php
index 31277fe808624e8896cbc96a46e90aeb340dfb82..963bddb97f05adee663b72dd52cdb4c82a4346f6 100644
--- a/core/modules/filter/src/Plugin/Filter/FilterAutoP.php
+++ b/core/modules/filter/src/Plugin/Filter/FilterAutoP.php
@@ -60,7 +60,7 @@ public function process($text, $langcode) {
       }
       elseif (!$ignore) {
         // Skip if the next chunk starts with Twig theme debug.
-        // @see twig_render_template()
+        // @see \Drupal\Core\Template\TwigThemeEngine::renderTemplate()
         if (isset($chunks[$i + 1]) && $chunks[$i + 1] === '<!-- THEME DEBUG -->') {
           $chunk = rtrim($chunk, "\n");
           $output .= $chunk;
@@ -102,9 +102,16 @@ public function process($text, $langcode) {
         $chunk = preg_replace('!<p>\s*(</?' . $block . '[^>]*>)!', "$1", $chunk);
         $chunk = preg_replace('!(</?' . $block . '[^>]*>)\s*</p>!', "$1", $chunk);
         // Make line breaks.
-        $chunk = preg_replace('|(?<!<br />)\s*\n|', "<br />\n", $chunk);
-        $chunk = preg_replace('!(</?' . $block . '[^>]*>)\s*<br />!', "$1", $chunk);
-        $chunk = preg_replace('!<br />(\s*</?(?:p|li|div|dl|dd|dt|th|pre|td|ul|ol)>)!', '$1', $chunk);
+        $chunk = preg_replace_callback('|(<br[^>]*>)?\s*\n|', function ($matches) {
+          // Preserve existing break tags with attributes.
+          if (isset($matches[1])) {
+            return $matches[1] . "\n";
+          }
+          // Otherwise create new break tag.
+          return "<br />\n";
+        }, $chunk);
+        $chunk = preg_replace('!(</?' . $block . '[^>]*>)\s*<br[^>]*>!', "$1", $chunk);
+        $chunk = preg_replace('!<br[^>]*>(\s*</?(?:p|li|div|dl|dd|dt|th|pre|td|ul|ol)>)!', '$1', $chunk);
         $chunk = preg_replace('/&([^#])(?![A-Za-z0-9]{1,8};)/', '&amp;$1', $chunk);
       }
       $output .= $chunk;
diff --git a/core/modules/filter/tests/src/Kernel/FilterKernelTest.php b/core/modules/filter/tests/src/Kernel/FilterKernelTest.php
index f7d12011b6793145a4e0dcb8087638b629e40889..e6569d5e71a21c966bd7809be03880e34db495b1 100644
--- a/core/modules/filter/tests/src/Kernel/FilterKernelTest.php
+++ b/core/modules/filter/tests/src/Kernel/FilterKernelTest.php
@@ -490,6 +490,15 @@ public function testLineBreakFilter(): void {
       '<drupal-media data-caption=" " data-entity-type="media" data-entity-uuid="dbb16f97-cd11-4357-acde-cd09e19e312b"></drupal-media>' => [
         '<p><drupal-media data-caption=" " data-entity-type="media" data-entity-uuid="dbb16f97-cd11-4357-acde-cd09e19e312b"></drupal-media></p>' => FALSE,
       ],
+      // Bug 3083786, duplicate br tags after a line ending in a break tag that
+      // doesn't exactly match <br />
+      'aaa<br>
+      bbb<br class="ccc" />
+      ddd<br/>' => [
+        "<p>aaa<br>\n" => TRUE,
+        "bbb<br class=\"ccc\" />\n" => TRUE,
+        "ddd</p>" => TRUE,
+      ],
     ];
     // cSpell:enable
     $this->assertFilteredString($filter, $tests);
