diff --git a/core/core.services.yml b/core/core.services.yml
index 31b7a5b6aa..1287b3a2fa 100644
--- a/core/core.services.yml
+++ b/core/core.services.yml
@@ -822,7 +822,7 @@ services:
   Drupal\Core\Menu\MenuLinkManagerInterface: '@plugin.manager.menu.link'
   menu.link_tree:
     class: Drupal\Core\Menu\MenuLinkTree
-    arguments: ['@menu.tree_storage', '@plugin.manager.menu.link', '@router.route_provider', '@menu.active_trail', '@callable_resolver']
+    arguments: ['@menu.tree_storage', '@plugin.manager.menu.link', '@router.route_provider', '@menu.active_trail', '@callable_resolver', '@event_dispatcher']
   Drupal\Core\Menu\MenuLinkTreeInterface: '@menu.link_tree'
   menu.default_tree_manipulators:
     class: Drupal\Core\Menu\DefaultMenuLinkTreeManipulators
diff --git a/core/lib/Drupal/Core/Menu/MenuLinkBase.php b/core/lib/Drupal/Core/Menu/MenuLinkBase.php
index 7757f45159..6ef381f6e1 100644
--- a/core/lib/Drupal/Core/Menu/MenuLinkBase.php
+++ b/core/lib/Drupal/Core/Menu/MenuLinkBase.php
@@ -3,7 +3,7 @@
 namespace Drupal\Core\Menu;

 use Drupal\Component\Plugin\Exception\PluginException;
-use Drupal\Core\Cache\Cache;
+use Drupal\Core\Cache\RefinableCacheableDependencyTrait;
 use Drupal\Core\Plugin\PluginBase;
 use Drupal\Core\Url;

@@ -12,6 +12,8 @@ use Drupal\Core\Url;
  */
 abstract class MenuLinkBase extends PluginBase implements MenuLinkInterface {

+  use RefinableCacheableDependencyTrait;
+
   /**
    * The list of definition values where an override is allowed.
    *
@@ -209,25 +211,4 @@ abstract class MenuLinkBase extends PluginBase implements MenuLinkInterface {
     throw new PluginException("Menu link plugin with ID '{$this->getPluginId()}' does not support deletion");
   }

-  /**
-   * {@inheritdoc}
-   */
-  public function getCacheMaxAge() {
-    return Cache::PERMANENT;
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  public function getCacheContexts() {
-    return [];
-  }
-
-  /**
-   * {@inheritdoc}
-   */
-  public function getCacheTags() {
-    return [];
-  }
-
 }
diff --git a/core/lib/Drupal/Core/Menu/MenuLinkInterface.php bcore//lib/Drupal/Core/Menu/MenuLinkInterface.php
index 0699aef361..4a420627fc 100644
--- a/core/lib/Drupal/Core/Menu/MenuLinkInterface.php
+++ b/core/lib/Drupal/Core/Menu/MenuLinkInterface.php
@@ -4,13 +4,13 @@ namespace Drupal\Core\Menu;

 use Drupal\Component\Plugin\PluginInspectionInterface;
 use Drupal\Component\Plugin\DerivativeInspectionInterface;
-use Drupal\Core\Cache\CacheableDependencyInterface;
+use Drupal\Core\Cache\RefinableCacheableDependencyInterface;
 use Drupal\Core\Url;

 /**
  * Defines an interface for classes providing a type of menu link.
  */
-interface MenuLinkInterface extends PluginInspectionInterface, DerivativeInspectionInterface, CacheableDependencyInterface {
+interface MenuLinkInterface extends PluginInspectionInterface, DerivativeInspectionInterface, RefinableCacheableDependencyInterface {

   /**
    * Returns the weight of the menu link.
diff --git a/core/lib/Drupal/Core/Menu/MenuLinkTree.php b/core/lib/Drupal/Core/Menu/MenuLinkTree.php
index 0d2035055b..45049f0138 100644
--- a/core/lib/Drupal/Core/Menu/MenuLinkTree.php
+++ b/core/lib/Drupal/Core/Menu/MenuLinkTree.php
@@ -9,6 +9,7 @@ use Drupal\Core\Routing\PreloadableRouteProviderInterface;
 use Drupal\Core\Routing\RouteProviderInterface;
 use Drupal\Core\Template\Attribute;
 use Drupal\Core\Utility\CallableResolver;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;

 /**
  * Implements the loading, transforming and rendering of menu link trees.
@@ -28,6 +29,8 @@ class MenuLinkTree implements MenuLinkTreeInterface {
    *   The active menu trail service.
    * @param \Drupal\Core\Utility\CallableResolver $callableResolver
    *   The callable resolver.
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface|null $eventDispatcher
+   *   The event dispatcher.
    */
   public function __construct(
     protected MenuTreeStorageInterface $treeStorage,
@@ -35,7 +38,12 @@ class MenuLinkTree implements MenuLinkTreeInterface {
     protected RouteProviderInterface $routeProvider,
     protected MenuActiveTrailInterface $menuActiveTrail,
     protected CallableResolver $callableResolver,
+    protected ?EventDispatcherInterface $eventDispatcher = NULL,
   ) {
+    if ($eventDispatcher === NULL) {
+      @trigger_error('Calling ' . __METHOD__ . '() without the $eventDispatcher argument is deprecated in drupal:11.4.0 and will be required in drupal:12.0.0. See https://www.drupal.org/node/3463907', E_USER_DEPRECATED);
+      $this->eventDispatcher = \Drupal::service('event_dispatcher');
+    }
   }

   /**
@@ -102,7 +110,9 @@ class MenuLinkTree implements MenuLinkTreeInterface {
    * {@inheritdoc}
    */
   public function transform(array $tree, array $manipulators) {
-    foreach ($manipulators as $manipulator) {
+    $event = new MenuLinkTreeManipulatorsAlterEvent($tree, $manipulators, $this);
+    $this->eventDispatcher->dispatch($event);
+    foreach ($event->getManipulators() as $manipulator) {
       $callable = $this->callableResolver->getCallableFromDefinition($manipulator['callable']);
       // Prepare the arguments for the menu tree manipulator callable; the first
       // argument is always the menu link tree.
diff --git a/core/lib/Drupal/Core/Menu/MenuLinkTreeManipulatorsAlterEvent.php bcore//lib/Drupal/Core/Menu/MenuLinkTreeManipulatorsAlterEvent.php
new file mode 100644
index 0000000000..d659b93326
--- /dev/null
+++ b/core/lib/Drupal/Core/Menu/MenuLinkTreeManipulatorsAlterEvent.php
@@ -0,0 +1,61 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Drupal\Core\Menu;
+
+use Drupal\Component\EventDispatcher\Event;
+
+/**
+ * Represents menu link tree manipulator information as event.
+ */
+class MenuLinkTreeManipulatorsAlterEvent extends Event {
+
+  public function __construct(
+    protected array $tree,
+    protected array $manipulators,
+    protected MenuLinkTreeInterface $menuLinkTree,
+  ) {
+  }
+
+  /**
+   * The MenuLinkElement tree.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeElement[]
+   *   The menu tree.
+   */
+  public function getTree(): array {
+    return $this->tree;
+  }
+
+  /**
+   * Get the manipulators for the tree.
+   *
+   * @return array
+   *   The menu tree manipulators.
+   */
+  public function getManipulators(): array {
+    return $this->manipulators;
+  }
+
+  /**
+   * Set the manipulators for the tree.
+   *
+   * @param array $manipulators
+   *   The menu tree manipulators.
+   */
+  public function setManipulators(array $manipulators): void {
+    $this->manipulators = $manipulators;
+  }
+
+  /**
+   * The MenuLinkTree.
+   *
+   * @return \Drupal\Core\Menu\MenuLinkTreeInterface
+   *   The menu link tree.
+   */
+  public function getMenuLinkTree(): MenuLinkTreeInterface {
+    return $this->menuLinkTree;
+  }
+
+}
diff --git a/core/modules/navigation/navigation.services.yml b/core/modules/navigation/navigation.services.yml
index 925efe58a4..3d6df4fd4e 100644
--- a/core/modules/navigation/navigation.services.yml
+++ b/core/modules/navigation/navigation.services.yml
@@ -29,6 +29,7 @@ services:
         '@menu.active_trail',
         '@callable_resolver',
         '@module_handler',
+        '@event_dispatcher',
       ]

   navigation.entity_route_helper:
diff --git a/core/modules/navigation/src/Menu/NavigationMenuLinkTree.php bcore//modules/navigation/src/Menu/NavigationMenuLinkTree.php
index 68ce42fc02..cd0adc4ce8 100644
--- a/core/modules/navigation/src/Menu/NavigationMenuLinkTree.php
+++ b/core/modules/navigation/src/Menu/NavigationMenuLinkTree.php
@@ -13,6 +13,7 @@ use Drupal\Core\Menu\MenuLinkTree;
 use Drupal\Core\Menu\MenuTreeStorageInterface;
 use Drupal\Core\Routing\RouteProviderInterface;
 use Drupal\Core\Utility\CallableResolver;
+use Symfony\Component\EventDispatcher\EventDispatcherInterface;

 /**
  * Extends MenuLinkTree to add specific theme suggestions for the navigation.
@@ -36,6 +37,8 @@ final class NavigationMenuLinkTree extends MenuLinkTree {
    *   The callable resolver.
    * @param \Drupal\Core\Extension\ModuleHandlerInterface $moduleHandler
    *   The module handler.
+   * @param \Symfony\Component\EventDispatcher\EventDispatcherInterface|null $eventDispatcher
+   *   The event dispatcher.
    */
   public function __construct(
     MenuTreeStorageInterface $treeStorage,
@@ -44,8 +47,13 @@ final class NavigationMenuLinkTree extends MenuLinkTree {
     MenuActiveTrailInterface $menuActiveTrail,
     CallableResolver $callableResolver,
     protected ModuleHandlerInterface $moduleHandler,
+    protected ?EventDispatcherInterface $eventDispatcher = NULL,
   ) {
-    parent::__construct($treeStorage, $menuLinkManager, $routeProvider, $menuActiveTrail, $callableResolver);
+    if ($eventDispatcher === NULL) {
+      @trigger_error('Calling ' . __METHOD__ . '() without the $eventDispatcher argument is deprecated in drupal:11.4.0 and will be required in drupal:12.0.0. See https://www.drupal.org/node/3463907', E_USER_DEPRECATED);
+      $eventDispatcher = \Drupal::service('event_dispatcher');
+    }
+    parent::__construct($treeStorage, $menuLinkManager, $routeProvider, $menuActiveTrail, $callableResolver, $eventDispatcher);
   }

   /**
diff --git a/core/modules/system/tests/modules/menu_test/menu_test.routing.yml bcore//modules/system/tests/modules/menu_test/menu_test.routing.yml
index d9f611164c..622850104a 100644
--- a/core/modules/system/tests/modules/menu_test/menu_test.routing.yml
+++ b/core/modules/system/tests/modules/menu_test/menu_test.routing.yml
@@ -585,6 +585,30 @@ menu_test.breadcrumb3:
   requirements:
     _access: 'TRUE'

+menu_test.manipulator1:
+  path: '/menu-test/manipulators/1'
+  defaults:
+    _controller: '\Drupal\menu_test\Controller\MenuTestController::menuTestCallback'
+    _title: 'Link 1'
+  requirements:
+    _access: 'TRUE'
+
+menu_test.manipulator2:
+  path: '/menu-test/manipulators/2'
+  defaults:
+    _controller: '\Drupal\menu_test\Controller\MenuTestController::menuTestCallback'
+    _title: 'Link 2'
+  requirements:
+    _access: 'TRUE'
+
+menu_test.manipulator3:
+  path: '/menu-test/manipulators/3'
+  defaults:
+    _controller: '\Drupal\menu_test\Controller\MenuTestController::menuTestCallback'
+    _title: 'Link 1'
+  requirements:
+    _access: 'TRUE'
+
 menu_test.skippable-breadcrumb:
   path: '/menu-test/skippable-breadcrumb'
   defaults:
diff --git a/core/modules/system/tests/modules/menu_test/menu_test.services.yml bcore//modules/system/tests/modules/menu_test/menu_test.services.yml
index e1a0aa059b..759dd1a5b5 100644
--- a/core/modules/system/tests/modules/menu_test/menu_test.services.yml
+++ b/core/modules/system/tests/modules/menu_test/menu_test.services.yml
@@ -9,3 +9,8 @@ services:
     arguments: ['@request_stack']
     tags:
       - { name: access_check, applies_to: _menu_test_session_access }
+
+  menu_test.event_subscriber.menu_link_tree:
+    class: Drupal\menu_test\EventSubscriber\MenuLinkTreeEventSubscriber
+    tags:
+      - { name: event_subscriber }
diff --git a/core/modules/system/tests/src/Unit/Menu/MenuLinkTreeTest.php bcore//modules/system/tests/src/Unit/Menu/MenuLinkTreeTest.php
index 51a9bd0e38..35b036c4b9 100644
--- a/core/modules/system/tests/src/Unit/Menu/MenuLinkTreeTest.php
+++ b/core/modules/system/tests/src/Unit/Menu/MenuLinkTreeTest.php
@@ -39,7 +39,8 @@ class MenuLinkTreeTest extends UnitTestCase {
       $this->createMock('\Drupal\Core\Menu\MenuLinkManagerInterface'),
       $this->createMock('\Drupal\Core\Routing\RouteProviderInterface'),
       $this->createMock('\Drupal\Core\Menu\MenuActiveTrailInterface'),
-      $this->createMock(CallableResolver::class)
+      $this->createMock(CallableResolver::class),
+      $this->createMock('\Symfony\Component\EventDispatcher\EventDispatcherInterface')
     );

     $cache_contexts_manager = $this->getMockBuilder('Drupal\Core\Cache\Context\CacheContextsManager')
diff --git a/core/modules/toolbar/toolbar.services.yml bcore//modules/toolbar/toolbar.services.yml
index 67a1b07191..db08c0e763 100644
--- a/core/modules/toolbar/toolbar.services.yml
+++ b/core/modules/toolbar/toolbar.services.yml
@@ -16,4 +16,4 @@ services:
       - { name: page_cache_request_policy }
   toolbar.menu_tree:
     class: Drupal\toolbar\Menu\ToolbarMenuLinkTree
-    arguments: ['@menu.tree_storage', '@plugin.manager.menu.link', '@router.route_provider', '@menu.active_trail', '@callable_resolver']
+    arguments: ['@menu.tree_storage', '@plugin.manager.menu.link', '@router.route_provider', '@menu.active_trail', '@callable_resolver', '@event_dispatcher']
diff --git a/core/tests/Drupal/KernelTests/Core/Menu/MenuLinkTreeTest.php bcore//tests/Drupal/KernelTests/Core/Menu/MenuLinkTreeTest.php
index ee7b5c02d2..f3d67e57e4 100644
--- a/core/tests/Drupal/KernelTests/Core/Menu/MenuLinkTreeTest.php
+++ b/core/tests/Drupal/KernelTests/Core/Menu/MenuLinkTreeTest.php
@@ -174,4 +174,49 @@ class MenuLinkTreeTest extends KernelTestBase {
     $this->assertSame(['user_logout_example'], $get_accessible_links());
   }

+  /**
+   * Tests menu link manipulator alter events.
+   */
+  public function testMenuManipulatorAlter(): void {
+    $links = [
+      1 => MenuLinkMock::createMock([
+        'id' => 'test.link1',
+        'route_name' => 'menu_test.manipulator1',
+        'title' => 'Link 1',
+        'parent' => '',
+      ]),
+      2 => MenuLinkMock::createMock([
+        'id' => 'test.link2',
+        'route_name' => 'menu_test.manipulator2',
+        'title' => 'Link 2',
+        'parent' => '',
+      ]),
+      3 => MenuLinkMock::createMock([
+        'id' => 'test.link3',
+        'route_name' => 'menu_test.manipulator3',
+        'title' => 'Link 3',
+        'parent' => '',
+      ]),
+    ];
+    foreach ($links as $instance) {
+      $this->menuLinkManager->addDefinition($instance->getPluginId(),
+        $instance->getPluginDefinition());
+    }
+    $parameters = new MenuTreeParameters();
+    $tree = $this->linkTree->load('mock', $parameters);
+    $manipulators = [
+      ['callable' => 'menu.default_tree_manipulators:generateIndexAndSort'],
+    ];
+    $tree = $this->linkTree->transform($tree, $manipulators);
+
+    $this->assertCount(3, $tree);
+    /** @var \Drupal\Core\Menu\MenuLinkTreeElement[] $tree */
+    list($link1, $link2, $link3) = array_values($tree);
+
+    // Check that the correct links have an additional class.
+    $this->assertNotContains('menu-test-link', $link1->options['attributes']['class'] ?? []);
+    $this->assertContains('menu-test-link', $link2->options['attributes']['class'] ?? []);
+    $this->assertContains('menu-test-link', $link3->options['attributes']['class'] ?? []);
+  }
+
 }
