diff --git a/core/lib/Drupal/Core/Template/TwigThemeEngine.php b/core/lib/Drupal/Core/Template/TwigThemeEngine.php
index 23017a967c1..c40b5cd57eb 100644
--- a/core/lib/Drupal/Core/Template/TwigThemeEngine.php
+++ b/core/lib/Drupal/Core/Template/TwigThemeEngine.php
@@ -52,8 +52,8 @@ public function renderTemplate(string $template_file, array $variables): string|
         'debug_suffix' => '',
       ];
 
-      $output['debug_prefix'] .= "\n\n<!-- THEME DEBUG -->";
-      $output['debug_prefix'] .= "\n<!-- THEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
+      $output['debug_prefix'] .= "<!--\n\nTHEME DEBUG -->";
+      $output['debug_prefix'] .= "<!--\nTHEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
       // If there are theme suggestions, reverse the array so more specific
       // suggestions are shown first.
       if (!empty($variables['theme_hook_suggestions'])) {
@@ -100,10 +100,10 @@ public function renderTemplate(string $template_file, array $variables): string|
           }
           $suggestion = $prefix . ' ' . $template;
         }
-        $output['debug_info'] .= "\n<!-- FILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";
+        $output['debug_info'] .= "<!--\nFILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";
 
         if (!empty($invalid_suggestions)) {
-          $output['debug_info'] .= "\n<!-- INVALID FILE NAME SUGGESTIONS:";
+          $output['debug_info'] .= "<!--\nINVALID FILE NAME SUGGESTIONS:";
           $output['debug_info'] .= "\n   See https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Render!theme.api.php/function/hook_theme_suggestions_alter";
           $output['debug_info'] .= "\n   " . Html::escape(implode("\n   ", $invalid_suggestions));
           $output['debug_info'] .= "\n-->";
@@ -116,8 +116,8 @@ public function renderTemplate(string $template_file, array $variables): string|
         $template_override_status_output = "💡 BEGIN CUSTOM TEMPLATE OUTPUT";
         $template_override_suffix_output = "END CUSTOM TEMPLATE OUTPUT";
       }
-      $output['debug_info']   .= "\n<!-- " . $template_override_status_output . " from '" . Html::escape($template_file) . "' -->\n";
-      $output['debug_suffix'] .= "\n<!-- " . $template_override_suffix_output . " from '" . Html::escape($template_file) . "' -->\n\n";
+      $output['debug_info']   .= "<!--\n" . $template_override_status_output . " from '" . Html::escape($template_file) . "'\n-->";
+      $output['debug_suffix'] .= "<!--\n" . $template_override_suffix_output . " from '" . Html::escape($template_file) . "'\n\n-->";
       // This output has already been rendered and is therefore considered safe.
       return Markup::create(implode('', $output));
     }
diff --git a/core/modules/filter/filter.module b/core/modules/filter/filter.module
index 9eec4e2e549..414214e348a 100644
--- a/core/modules/filter/filter.module
+++ b/core/modules/filter/filter.module
@@ -659,8 +659,7 @@ function _filter_autop($text) {
     elseif (!$ignore) {
       // Skip if the next chunk starts with Twig theme debug.
       // @see twig_render_template()
-      if (isset($chunks[$i + 1]) && $chunks[$i + 1] === '<!-- THEME DEBUG -->') {
-        $chunk = rtrim($chunk, "\n");
+      if (isset($chunks[$i + 1]) && $chunks[$i + 1] === "<!--\n\nTHEME DEBUG -->") {
         $output .= $chunk;
         continue;
       }
@@ -669,10 +668,9 @@ function _filter_autop($text) {
       // @see twig_render_template()
       if (isset($chunks[$i - 1])) {
         if (
-          str_starts_with($chunks[$i - 1], '<!-- BEGIN OUTPUT from ')
-          || str_starts_with($chunks[$i - 1], '<!-- 💡 BEGIN CUSTOM TEMPLATE OUTPUT from ')
+          str_starts_with($chunks[$i - 1], "<!--\nBEGIN OUTPUT from ")
+          || str_starts_with($chunks[$i - 1], "<!--\n💡 BEGIN CUSTOM TEMPLATE OUTPUT from ")
         ) {
-          $chunk = ltrim($chunk, "\n");
           $output .= $chunk;
           continue;
         }
