diff --git a/core/lib/Drupal/Core/Template/TwigThemeEngine.php b/core/lib/Drupal/Core/Template/TwigThemeEngine.php
index 23017a967c..4089ea3c23 100644
--- a/core/lib/Drupal/Core/Template/TwigThemeEngine.php
+++ b/core/lib/Drupal/Core/Template/TwigThemeEngine.php
@@ -52,8 +52,8 @@ class TwigThemeEngine implements ThemeEngineInterface {
         'debug_suffix' => '',
       ];

-      $output['debug_prefix'] .= "\n\n<!-- THEME DEBUG -->";
-      $output['debug_prefix'] .= "\n<!-- THEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
+      $output['debug_prefix'] .= "<!-- THEME DEBUG -->";
+      $output['debug_prefix'] .= "<!-- THEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
       // If there are theme suggestions, reverse the array so more specific
       // suggestions are shown first.
       if (!empty($variables['theme_hook_suggestions'])) {
@@ -100,7 +100,7 @@ class TwigThemeEngine implements ThemeEngineInterface {
           }
           $suggestion = $prefix . ' ' . $template;
         }
-        $output['debug_info'] .= "\n<!-- FILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";
+        $output['debug_info'] .= "<!-- FILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";

         if (!empty($invalid_suggestions)) {
           $output['debug_info'] .= "\n<!-- INVALID FILE NAME SUGGESTIONS:";
@@ -116,8 +116,8 @@ class TwigThemeEngine implements ThemeEngineInterface {
         $template_override_status_output = "💡 BEGIN CUSTOM TEMPLATE OUTPUT";
         $template_override_suffix_output = "END CUSTOM TEMPLATE OUTPUT";
       }
-      $output['debug_info']   .= "\n<!-- " . $template_override_status_output . " from '" . Html::escape($template_file) . "' -->\n";
-      $output['debug_suffix'] .= "\n<!-- " . $template_override_suffix_output . " from '" . Html::escape($template_file) . "' -->\n\n";
+      $output['debug_info']   .= "<!-- " . $template_override_status_output . " from '" . Html::escape($template_file) . "' -->\n";
+      $output['debug_suffix'] .= "<!-- " . $template_override_suffix_output . " from '" . Html::escape($template_file) . "' -->";
       // This output has already been rendered and is therefore considered safe.
       return Markup::create(implode('', $output));
     }
diff --git a/core/themes/engines/twig/twig.engine b/core/themes/engines/twig/twig.engine
index 4edf978c8c..c6714aeba6 100644
--- a/core/themes/engines/twig/twig.engine
+++ b/core/themes/engines/twig/twig.engine
@@ -54,8 +54,8 @@ function twig_render_template($template_file, array $variables): string|MarkupIn
       'debug_suffix' => '',
     ];

-    $output['debug_prefix'] .= "\n\n<!-- THEME DEBUG -->";
-    $output['debug_prefix'] .= "\n<!-- THEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
+    $output['debug_prefix'] .= "<!-- THEME DEBUG -->";
+    $output['debug_prefix'] .= "<!-- THEME HOOK: '" . Html::escape($variables['theme_hook_original']) . "' -->";
     // If there are theme suggestions, reverse the array so more specific
     // suggestions are shown first.
     if (!empty($variables['theme_hook_suggestions'])) {
@@ -103,10 +103,10 @@ function twig_render_template($template_file, array $variables): string|MarkupIn
         }
         $suggestion = $prefix . ' ' . $template;
       }
-      $output['debug_info'] .= "\n<!-- FILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";
+      $output['debug_info'] .= "<!-- FILE NAME SUGGESTIONS:\n   " . Html::escape(implode("\n   ", $suggestions)) . "\n-->";

       if (!empty($invalid_suggestions)) {
-        $output['debug_info'] .= "\n<!-- INVALID FILE NAME SUGGESTIONS:";
+        $output['debug_info'] .= "<!-- INVALID FILE NAME SUGGESTIONS:";
         $output['debug_info'] .= "\n   See https://api.drupal.org/api/drupal/core!lib!Drupal!Core!Render!theme.api.php/function/hook_theme_suggestions_alter";
         $output['debug_info'] .= "\n   " . Html::escape(implode("\n   ", $invalid_suggestions));
         $output['debug_info'] .= "\n-->";
@@ -119,8 +119,8 @@ function twig_render_template($template_file, array $variables): string|MarkupIn
       $template_override_status_output = "💡 BEGIN CUSTOM TEMPLATE OUTPUT";
       $template_override_suffix_output = "END CUSTOM TEMPLATE OUTPUT";
     }
-    $output['debug_info']   .= "\n<!-- " . $template_override_status_output . " from '" . Html::escape($template_file) . "' -->\n";
-    $output['debug_suffix'] .= "\n<!-- " . $template_override_suffix_output . " from '" . Html::escape($template_file) . "' -->\n\n";
+    $output['debug_info']   .= "<!-- " . $template_override_status_output . " from '" . Html::escape($template_file) . "' -->\n";
+    $output['debug_suffix'] .= "<!-- " . $template_override_suffix_output . " from '" . Html::escape($template_file) . "' -->\n\n";
     // This output has already been rendered and is therefore considered safe.
     return Markup::create(implode('', $output));
   }
