diff --git a/lib/Drupal/Core/Config/StorageComparer.php b/lib/Drupal/Core/Config/StorageComparer.php
index 820fe041fc..3323c6d65e 100644
--- a/lib/Drupal/Core/Config/StorageComparer.php
+++ b/lib/Drupal/Core/Config/StorageComparer.php
@@ -299,6 +299,15 @@ class StorageComparer implements StorageComparerInterface {
     foreach (array_intersect($this->sourceNames[$collection], $this->targetNames[$collection]) as $name) {
       $source_data = $this->getSourceStorage($collection)->read($name);
       $target_data = $this->getTargetStorage($collection)->read($name);
+
+      // Sort array keys to ensure only actual configuration changes are
+      // detected. This does not affect the export process, as all
+      // configuration is always exported in full.
+      if (is_array($source_data) && is_array($target_data)) {
+        $this->ksortRecursive($source_data);
+        $this->ksortRecursive($target_data);
+      }
+
       if ($source_data !== $target_data) {
         if (isset($source_data['uuid']) && $source_data['uuid'] !== $target_data['uuid']) {
           // The entity has the same file as an existing entity but the UUIDs do
@@ -321,6 +330,22 @@ class StorageComparer implements StorageComparerInterface {
     }
   }
 
+  /**
+   * Sorts an array recursively based on the keys.
+   *
+   * @param array $array
+   *   Array to sort.
+   */
+  protected function ksortRecursive(array &$array): void {
+    ksort($array);
+    foreach ($array as &$value) {
+      if (is_array($value)) {
+        $this->ksortRecursive($value);
+      }
+    }
+    unset($value);
+  }
+
   /**
    * Creates the rename changelist.
    *
