diff --git a/editoria11y.module b/editoria11y.module
index abcf866144bdefd3cf8e14692612eec3528072a0..2c3e3f538150881858385f07f72833ed0718719e 100644
--- a/editoria11y.module
+++ b/editoria11y.module
@@ -280,6 +280,9 @@ function editoria11y_page_attachments(array &$attachments): void {
   $attachments['#attached']['drupalSettings']['editoria11y']['dismissals'] = $dismissals;
   $attachments['#attached']['drupalSettings']['editoria11y']['theme'] = $config->get('ed11y_theme');
   $attachments['#attached']['drupalSettings']['editoria11y']['custom_tests'] = $config->get('custom_tests');
-  $attachments['#attached']['drupalSettings']['editoria11y']['css_url'] = base_path() . Drupal::service('extension.list.module')->getPath('editoria11y');
+  /** @var \Drupal\Core\File\FileUrlGeneratorInterface $fileUrlGenerator */
+  $fileUrlGenerator = \Drupal::service('file_url_generator');
+  $attachments['#attached']['drupalSettings']['editoria11y']['css_url'] =  $fileUrlGenerator
+    ->generateString(Drupal::service('extension.list.module')->getPath('editoria11y'));

 }
