diff --git a/src/LinkIterator.php b/src/LinkIterator.php
index 3c1fda0..76ad3cc 100644
--- a/src/LinkIterator.php
+++ b/src/LinkIterator.php
@@ -120,7 +120,7 @@ class LinkIterator implements LinkIteratorInterface {
       $overridden_options = $item['url']->getOptions();
       $options += $overridden_options;
       if (isset($options['attributes'])) {
-        $overridden_options['attributes'] = array_merge_recursive($options['attributes'], $overridden_options['attributes']);
+        $overridden_options['attributes'] = array_merge_recursive($options['attributes'], $overridden_options['attributes'] ?? []);
       }
       $item['url']->setOptions($options);
     }
