diff --git a/src/ViewUnpublishedPermissions.php b/src/ViewUnpublishedPermissions.php
index de646f4..9ae30ab 100644
--- a/src/ViewUnpublishedPermissions.php
+++ b/src/ViewUnpublishedPermissions.php
@@ -46,6 +46,11 @@ class ViewUnpublishedPermissions {
     return [
       "view any unpublished $type_id content" => [
         'title' => $this->t('%type_name: View any unpublished content', $type_params),
+        'dependencies' => [
+          $type->getConfigDependencyKey() => [
+            $type->getConfigDependencyName()
+          ],
+        ],
       ],
     ];
   }
